### PR TITLE
hle/applets: add AppletFactory

### PR DESCRIPTION
**Pull request by valentinvanelslande**
After this PR is merged, i will open a ErrEula PR (log error code and text for SDL frontend, show a dialog for Qt frontend, this class will be used)

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3620)

### CODE REVIEW

File: src/core/hle/applets/factory.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    ~AppletFactory() {
+        Clear();
+    }
+
+    void Clear() {
+        callbacks.clear();
+    }
+
+    void Register(const std::string& name, CallbackType callback) {
+        callbacks.emplace(std::move(name), std::move(callback));
```
Comment (lioncash) [inline src/core/hle/applets/factory.h]: std::move doesn't do anything here for `name` You likely want to take `name` by value, not as a const reference.

File: src/core/hle/applets/factory.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    ResultType Launch(const std::string& name, const ConfigType& config) {
+        auto it = callbacks.find(name);
+        if (it != callbacks.end())
+            return it->second(config);
+        return default_result;
+    }
+
+    ResultType default_result;
```
Comment (lioncash) [inline src/core/hle/applets/factory.h]: This shouldn't be public.

File: src/core/hle/applets/factory.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#include <map>
+#include <string>
+
+namespace HLE {
+namespace Applets {
+
+template <typename CallbackType, typename ResultType, typename ConfigType>
+class AppletFactory {
+public:
+    ~AppletFactory() {
```
Comment (lioncash) [inline src/core/hle/applets/factory.h]: The destructor isn't necessary, when the factory goes out of scope it'll automatically deallocate everything in the map.

File: src/core/hle/applets/factory.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <functional>
+#include <map>
+#include <string>
```
Comment (lioncash) [inline src/core/hle/applets/factory.h]: `<utility>` needs to be included for `std::move`

File: src/core/hle/applets/factory.h (hunk 2 above)
Comment (valentinvanelslande) [reply src/core/hle/applets/factory.h]: in the constructors of the factories, that will be changed

Comment (lioncash) [reply src/core/hle/applets/factory.h]: I don't understand what you mean. Can you elaborate? Regardless of the implementation, it shouldn't be a requirement for this to be public.

Comment (valentinvanelslande) [reply src/core/hle/applets/factory.h]: changed to protected, in the constructors of the factories that is changed

**Comment by wwylele** [comment]
Agree to what Subv said, this looks just like a generic simple wrapper of named function registration boilerplate, which can be easily achieved by just using the `std::(unordered_)map` directly, and which has no obvious special trait that deserves its own PR. Please open another PR with what you actually want to implement, which brings real value to the code base.

**Pull request closed without merging**